### PR TITLE
fix(generate): fix association alias duplication

### PR DESCRIPTION
**Pull request by arnaud-moncel**
## Pull Request checklist:

- [X] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [X] Create automatic tests
- [X] No automatic tests failures
- [X] Test manually the implemented changes
- [X] Review my own code (indentation, syntax, style, simplicity, readability)
- [ ] Wonder if you can improve the existing code

### CODE REVIEW

File: templates/app/models/sequelize-model.hbs [hunk 1]
```diff
@@ -28,18 +28,18 @@ module.exports = (sequelize, DataTypes) => {
...
       foreignKey: '{{reference.foreignKey}}',
-      otherKey: '{{reference.otherKey}}',{{#if reference.as}}
-      as: '{{reference.as}}',{{/if}}
+      otherKey: '{{reference.otherKey}}',
     {{else}}
       foreignKey: {
         name: '{{reference.foreignKeyName}}',
         field: '{{reference.foreignKey}}',
-      },{{#if reference.targetKey }}
-      target: {
```
Comment (arnaud-moncel) [inline templates/app/models/sequelize-model.hbs:38]: This option is not handled by Sequelize see the [documlentation](https://sequelize.org/master/class/lib/model.js~Model.html#static-method-belongsTo) about this.

File: services/dumper.js [hunk 2]
```diff
@@ -203,23 +203,11 @@ function Dumper(config) {
       };
     });
 
-    const referencesDefinition = references.map((reference) => {
-      const computedReference = {
-        ...reference,
-        isBelongsToMany: reference.association === 'belongsToMany',
-      };
-
-      if (reference.targetKey) {
```
Comment (arnaud-moncel) [inline services/dumper.js:212]: Not used since Sequelize doesn’t support `target` option.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 3]
```diff
@@ -139,6 +157,9 @@ function createReference(tableName, association, foreignKey, manyToManyForeignKe
   if (association === ASSOCIATION_TYPE_BELONGS_TO) {
     reference.ref = foreignKey.foreignTableName;
     reference.as = formatAliasName(foreignKey.columnName);
+    if (foreignKey.foreignColumnName !== 'id') {
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:160]: Linked to `target` option refactoring.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 4]
```diff
@@ -157,8 +178,10 @@ function createReference(tableName, association, foreignKey, manyToManyForeignKe
     reference.as = _.camelCase(formater(`${prefix}${foreignKey.tableName}`));
   }
 
-  if (foreignKey.foreignColumnName !== 'id') {
-    reference.targetKey = foreignKey.foreignColumnName;
+  if (isReferenceAlreadyExist(existingsReferences, reference)) return null;
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:181]: This is the fix of duplicate reference.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 5]
```diff
@@ -157,8 +178,10 @@ function createReference(tableName, association, foreignKey, manyToManyForeignKe
...
+
+  if (isAssociationNameAlreadyExist(existingsReferences, reference)) {
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:183]: This is the fix of duplicate reference alias.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 6]
```diff
@@ -124,11 +124,29 @@ function checkUnicity(primaryKeys, uniqueIndexes, columnName) {
       indexColumnName.length === 1 && indexColumnName.includes(columnName));
 
   const isPrimary = _.isEqual([columnName], primaryKeys);
-  return { isPrimary, isUnique };
+  return isPrimary || isUnique;
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:127]: This is a little refacto to reduce cognitive complexity of `createAllReferences` function.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 7]
```diff
@@ -174,6 +197,31 @@ async function analyzeTable(table, config) {
   };
 }
 
+function createBelongsToReference(referenceTable, tableReferences, constraint) {
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:200]: This function is created to reduce cognitive complexity of `createAllReferences`.

File: services/analyzer/sequelize-tables-analyzer.js [hunk 8]
```diff
@@ -216,37 +264,32 @@ function createAllReferences(databaseSchema, schemaGenerated) {
...
-            ASSOCIATION_TYPE_BELONGS_TO,
-            constraint,
-          ),
-        );
-      }
     });
   });
 
-  return references;
+  // remove null references
```
Comment (arnaud-moncel) [inline services/analyzer/sequelize-tables-analyzer.js:284]: This piece of code is to reduce cognitive complexity of the function.

**Comment by arnaudbesnier** [comment]
Task linked: [CU-88ng48 P1 - Lumber - Ensure that alias and field names are unique in a model to prevent admin backend crash on first start](https://app.clickup.com/t/88ng48)

File: test-fixtures/mssql/doubleref.sql [hunk 9]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE doubleref (
```
Comment (guillaumedeslandes) [inline test-fixtures/mssql/doubleref.sql:1]:
```suggestion
DROP TABLE IF EXISTS doubleref;

CREATE TABLE doubleref (
```

File: test-fixtures/mysql/doubleref.sql [hunk 10]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE doubleref (
```
Comment (guillaumedeslandes) [inline test-fixtures/mysql/doubleref.sql:1]:
```suggestion
DROP TABLE IF EXISTS doubleref;

CREATE TABLE doubleref (
```

File: test-fixtures/postgres/doubleref.sql [hunk 11]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE doubleref (
```
Comment (guillaumedeslandes) [inline test-fixtures/postgres/doubleref.sql:1]:
```suggestion
DROP TABLE IF EXISTS doubleref;

CREATE TABLE doubleref (
```

File: templates/app/models/sequelize-model.hbs [hunk 12]
```diff
@@ -28,18 +28,18 @@ module.exports = (sequelize, DataTypes) => {
...
-        name: '{{reference.targetKey}}',{{#if reference.targetKeyColumnUnconventional}}
-        field: '{{reference.targetKey}}',{{/if}}
-      },{{/if}}{{#if reference.as}}
-      as: '{{reference.as}}',{{/if}}
+      },
+    {{#if reference.targetKey}}
+      targetKey: '{{reference.targetKey}}',
+    {{/if}}
```
Comment (guillaumedeslandes) [inline templates/app/models/sequelize-model.hbs:39]: Shouldn't this `if` be indented more?

File: services/analyzer/sequelize-tables-analyzer.js (hunk 6 above)
Comment (guillaumedeslandes) [inline services/analyzer/sequelize-tables-analyzer.js:127]:
```suggestion

  return isPrimary || isUnique;
```

File: services/analyzer/sequelize-tables-analyzer.js [hunk 13]
```diff
@@ -124,11 +124,29 @@ function checkUnicity(primaryKeys, uniqueIndexes, columnName) {
...
-  return { isPrimary, isUnique };
+  return isPrimary || isUnique;
+}
+
+function isAssociationNameAlreadyExist(existingReferences, newReference) {
+  return existingReferences.some((reference) => reference.as === newReference.as);
+}
+
+function isReferenceAlreadyExist(existingReferences, newReference) {
+  return existingReferences.some((reference) => (
```
Comment (guillaumedeslandes) [inline services/analyzer/sequelize-tables-analyzer.js:135]: Not fond of these names. They combine `is` and a verb.
I'd use either
* `associationNameAlreadyExists` or `associationNameIsAlreadyDefined` or `associationNameIsAlreadyUsed`.
* `referenceAlreadyExists` or `referenceIsAlreadyDefined` or `referenceIsAlreadyUsed`

File: services/analyzer/sequelize-tables-analyzer.js [hunk 14]
```diff
@@ -124,11 +124,30 @@ function checkUnicity(primaryKeys, uniqueIndexes, columnName) {
...
+  return isPrimary || isUnique;
+}
+
+function associationNameAlreadyExists(existingReferences, newReference) {
+  return existingReferences.some((reference) => reference.as === newReference.as);
+}
+
+function referenceAlreadyExists(existingReferences, newReference) {
+  return existingReferences.some((reference) => (
+    reference.ref === newReference.ref
```
Comment (guillaumedeslandes) [inline services/analyzer/sequelize-tables-analyzer.js:137]: I got a `null` value for `reference` while testing. Not sure of the reason, but worth investigating.

**Review by guillaumedeslandes — CHANGES_REQUESTED** [review]
See comment on `null` parameter.

**Review by guillaumedeslandes — APPROVED** [review]

**Pull request merged**